### DISCUSSION TITLE
CLR interop and generics

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#104] [Discussions]
Hi Leppie,

Is it possible to define CLR classes and hierarchies or implement interfaces in Scheme? The IronScheme clr api seems oriented around making use of already defined objects, but not defining new classes or implementing interfaces. Have you any plans to support interop at the type/interface level?

Any advice about using composition of Scheme records wrapping CLR types from your experience or knowledge? For example using plain ol records or "tinytalk" objects to provide more idiomatic Scheme access to types that need to play nice with, e.g. serialization libraries  (e.g., to/from JSON, YAML, etc., or storage in an object repository)?

**Comment by bcalco** [comment]
To further clarify, I could really use some clear examples of the following in Scheme:

1. Calling static or class instance methods that require generic type parameters 
2. Creating instances of generic types
3. Reflection over generic types (to understand how to process type metadata in idiomatic IronScheme)
4. (If possible) Creating new CLR types and instantiating instances

The documentation of the clr namespace is terse, which, as someone who can talk himself out of all the choices on a multiple guess exam, I find in some cases too ambiguous to process. :) 

I am trying to wrap several .NET libraries of various provenance (C# and F#) and would like to make sure I'm doing it right.

Perhaps to knock all the birds out with one stone, a tutorial on wrapping a library in idiomatic IronScheme could be helpful. Something like LiteDB (which could also show off concrete C#->Scheme LINQ translations) or a console parsing library like CommandLine.  I'm kinda working on such a thing, but running into questions, so I post this clarified version of my question in the hopes @leppie or somebody can spare a few moments to assist in further explaining how clr interop works (and what the specific limitations are, e.g., can't create new CLR types).

**Comment by leppie** [comment]
No, defining consumable CLR like classes is not possible from Scheme. Given how easy it is to do them from another language, I have never really bothered. Scheme records are CLR classes though, but it would be hard to consume them from other languages directly.

That said, you can (or at least if I remember correctly), it is possible to create Scheme record inheriting from a CLR class (probably limited to POCO objects).

Cant comment on serialization. Scheme records are binary serializable (aka binary formatter, scream all you like ;p), but that was that was done.

Edit: That said, this would be nice feature if someone can do it properly, but I fell short. Again, much easier to define this in C# and referencing it.

**Reply by bcalco** [reply]
Well, the binary serialization may not be a problem for how I want to use these objects. LiteDB is mainly what I'm interested in supporting with this "use case" if you will. How can I better understand what records are under the hood?

**Comment by leppie** [comment]
Consumption of .NET is where IronScheme is much better, in contrast to being a 'producer'.

1. Yes, `clr-call` supports generic arguments. 
2. Yes `clr-new` supports generic types
3. Not exactly understanding the question.
4. Yeah, as said, unlikely (except thru inheritance maybe)

Now for discussing generic types in IronScheme. 

`clr-call` allows one to specify a generic type for both instance and parameters.

A normal non-generic call would be something like `(clr-call Object ToString instance)`.

For a generic type instance, this would be `(clr-call (Foo Bar) ToString instance)`.

Here is an example for creating a generic type instance:
```
> (import (ironscheme clr))
> (clr-using System.Collections.Generic)
> (define l (clr-new (List int)))
> l
#<clr-type System.Collections.Generic.List`1[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]] "System.Collections.Generic.List`1[System.Int32]">
> (clr-call (List int) Add l 1)
> (clr-call (List int) (Add int) l 1)
> (clr-call (List int) ToArray l)
#fx(1 1)
```

When you get to parameters, it is a bit more tricky. 
(disclaimer: I might not be 100% right on this below)

To indicate a generic type, you would call `(clr-call Foo (Bar (List int)) instance arg)`. 
Where this in C# would be `Foo::Bar(List<int>)`.  
But `(clr-call Foo Bar instance arg)` should just work too as the compiler will attempt the cast.

If it was defined as `Foo::Bar<T>(List<T>)`, and you want to say use an `int`.
Then you would call it as `(clr-call Foo (Bar #(int)) instance arg)`.
Or more explicitly `(clr-call Foo (Bar #(int) (List int)) instance arg)`. 

The vector after the method will do generic binding.

IronScheme does not support open types in generics. 

Lastly, another nice feature I have been considering is define generic typed procedures in IronScheme, but it is not quite as simple as consuming them.

**Reply by bcalco** [reply]
This is helpful, thanks! ... but I reserve the right to ask more questions as I get into this. :)

**Comment by leppie** [comment]
This library also attempts to do some heavy lifting using late-binding, but again, I dont recall how much it can do (looking now, most just dealing with overloads).

https://github.com/IronScheme/IronScheme/blob/master/IronScheme/IronScheme.Console/lib/ironscheme/clr/dynamic.sls

**Comment by Zambito1** [comment]
@leppie putting this here since it seems to be related: is there a way to consume static classes in IronScheme?

For example:

```scheme
(import (ironscheme clr))

(clr-using System.Collections.Generic)
(clr-using System.Linq)

(clr-call (IEnumerable Object) ToArray (list "test"))
;; Fails with:
;; Unhandled exception during evaluation:
;; &who: "clr-call"
;; &message: "member could not be resolved on type: IEnumerable`1"
;; &syntax:
;;   form: ('(IEnumerable Object) 'ToArray (g$list$17816$1pFKpj '"test"))
;;   subform: "ToArray"
(clr-call-static Enumerable ToArray (list "test"))
;; Fails with:
;; Unhandled exception during evaluation:
;; &who: "ExtractTypeInfo"
;; &message: "type not found"
;; &syntax:
;;   form: "Enumerable"
;;   subform: #f
```

**Comment by leppie** [comment]
Hi

Everything you had, but with:

```
(clr-reference System.Core)
(clr-using System)
(clr-static-call Enumerable (ToArray #(Object)) (list "test"))
```

**Reply by Zambito1** [reply]
Thanks!

**Comment by tribals** [comment]
As I agree that it *might* be useful to declare classes using C#, and then consume them in IronScheme, it immediately raises a question: then, how to bundle them together and ship to user?

IIUC, currently there is no way to AOT-compile Scheme sources to CLR "assembly", so the only way of shipping is "here is the sources, then you get that IronScheme.Console thing somehow, and run the sources". This is OK unless you need to ship those "classes written in different language".

I think this is why people actually want to write classes in Scheme, not in C# - then you still able to "just ship sources".

BTW, nice kind of interoperability is [implemented in Kawa](https://andrebask.github.io/kawa-website/doc/Objects-Classes-and-Modules.html#Defining-new-classes) - there, you can easily not only call host, but construct classes and pass them to host, almost without any additional syntax - "looks like you just call classes".

My point is that "writing classes in C#" requires a lot more things to set up - eg, dotnet SDK (and I'm still not sure you can "just compile" C# with it) *and* build process in place, but writing Scheme requires "just downloading" IronScheme...

**Reply by tribals** [reply]
Put it other way around, and yes - one can "embed" Scheme into their projects, where build system is in place, then - again, you can't AOT-compile Scheme, so all "integration" available is "write Scheme as string literals", which is no-option for any "real-world application".
